### PR TITLE
[SPARK-4382] Add locations parameter to Twitter Stream

### PR DESCRIPTION
**Pull request by viirya**
When we request Tweet stream, geo-location is one of the most important parameters. In addition to the track parameter, the locations parameter is widely used to ask for the Tweets falling within the requested bounding boxes. This PR adds the locations parameter to existing APIs.

### CODE REVIEW

**Comment by viirya** [comment]
Hi @srowen, I refactor the codes for the comments. If you have time to review it, that would be great. Thanks.

File: external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterInputDStream.scala [hunk 1]
```diff
@@ -85,9 +93,14 @@ class TwitterReceiver(
         }
       })
 
-      val query = new FilterQuery
-      if (filters.size > 0) {
-        query.track(filters.toArray)
+      if (filters.size > 0 || locations.size > 0) {
```
Comment (srowen) [inline external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterInputDStream.scala]: Trivial but can this be `.nonEmpty` instead of `.size > 0` around here? slightly more readable

File: external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala [hunk 2]
```diff
@@ -53,7 +67,7 @@ object TwitterUtils {
    * @param jssc   JavaStreamingContext object
    */
   def createStream(jssc: JavaStreamingContext): JavaReceiverInputDStream[Status] = {
-    createStream(jssc.ssc, None)
+    createStream(jssc.ssc, None)
```
Comment (srowen) [inline external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala]: There's a couple spurious changes adding whitespace at the ends of lines. Not really worth fixing for its own sake but check the diff to see if you can fix it if you make another commit.

**Comment by srowen** [comment]
I think this is a lot better without all the overloads, yes.

**Comment by viirya** [comment]
Fixed in new commit. Thanks.

**Comment by viirya** [comment]
@srowen Is it ok going to merge this feature?

**Comment by srowen** [comment]
@viirya OK by me but @tdas will need to weigh in.

**Comment by viirya** [comment]
@tdas Do you have time to look at this to see if it is ready to merge? Thanks.

**Comment by viirya** [comment]
@srowen Does @tdas have time to look at this? If no, maybe others can?

**Comment by viirya** [comment]
@tdas Can you have a quick look of this pr? Thanks!

**Comment by tdas** [comment]
Apologies, I have been really backlogged with review because of the
upcoming Spark 1.3 release. The feature merge window for that has closed,
so we can consider this for Spark 1.4.

Also, there has been multiple attempts to add geolocation filtering to
Spark Streaming. In order decide between the different approaches, we will
have to do a proper design discussion with a design doc.

TD

**Comment by viirya** [comment]
@tdas Ah, that's great. Thanks for let me know that!

**Comment by srowen** [comment]
Since the JIRA is a duplicate of 2 others, can I suggest we close this and try to converge on SPARK-2788? not suggesting its PR is the one to merge, but at least we can centralize the discussion in one place. See also https://github.com/apache/spark/pull/1717

**Comment by viirya** [comment]
Good for me. Let's close this pr.
